proto/p4/p4runtime.proto

```diff
@@ -416,10 +433,20 @@ message PacketOut {
   repeated PacketMetadata metadata = 2;
 }
 
+// Used by the controller to ack a DigestList message. To avoid flooding the
+// controller, the switch must not generate digest notifications for the same
+// data until a DigestLickAck message with the same list_id is received or the
```

Comment:
Even with these mechanisms in place to avoid flooding the controller (which sound like good ideas to have), there is no guarantee that the client will consume messages by any specific deadline, or with any specific rate.

Thus there could be situations in which a server wants to a send a PacketIn or DigestList message to the client, but either it is sent over a reliable transport like TCP, with a queue building up that can fill to capacity, or it is sent over an unreliable transport like UDP, and be dropped even though the server successfully sent it.  Which is gRPC more like?

Assuming for the moment is is more like TCP, is it already specified or recommended somewhere what the server should do when it has a message to send to the server and the send attempt fails (if a nonblocking API call, because local send buffer is full) or blocks (if a blocking API call)?  The only thing that makes sense to me is if the server discards the PacketIn packet, or the digest buffer, counts it, and moves on.

Reply:
I think the gRPC transport is reliable (HTTP2 channel), but lets you do whatever you want at the application-level based on whether you use the synchronous API (blocking) or the asynchronous one, in which case you have more freedom to implement the behavior that you want. Ultimately, if the channel is too busy, messages will be dropped by the server, either at the time of calling Write (gRPC stream send) or before that if there is an extra queue in the implementation.

Just like for PacketIn, DigestList drops are not critical IMO as the network will recover. However, I agree that we should count the number of drops by the server.

Reply:
Agreed that losses of these messages should be non-critical.  It seems like a good idea to mention that the server can drop these kinds of messages if these messages arrive from the data plane faster than the client consumes them, so people choosing what kinds of PacketIn and Digest messages they use realize it can happen, and hopefully choose some other mechanism if they believe that reliability is critical.

Reply:
Agreed.